sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,329 @@
+			c <- runResult{err: err}
+		}
+		totalCount += 1
+
+		// Every second (roughly) we print out an update
+		if time.Since(timeStart).Seconds() > float64(lastPrint) {
+			thisCount := totalCount - sumInts(perSecondCount)
+			perSecondCount = append(perSecondCount, thisCount)
+			_, err = fmt.Fprintf(
+				w,
+				"%s\t%s\t%.2f\t\n",
+				commaIze(thisCount),
+				commaIze(totalCount),
+				float64(sumInts(perSecondCount))/time.Since(timeStart).Seconds(),
+			)
+			if err != nil {
+				c <- runResult{err: err}
+			}
+			lastPrint = time.Since(timeStart).Seconds() + 1.0
+			w.Flush()
+		}
+	}
+
+	elapsed := time.Since(timeStart).Seconds()
+
+	// Stop the proxy now
+	err := stop(p.GetMetadata().Name, nil)
+	if err != nil {
+		c <- runResult{err: err}
+	}
+	setRecordingMode("live")
+	c <- runResult{count: totalCount, timeInSeconds: elapsed, err: nil}
+}
+
+// runCleanup takes care of the semantics for tearing down a single iteration of a performance test.
+func runCleanup(p PerfTest) error {
+	err := p.Cleanup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runGlobalCleanup(p PerfTest) error {
+	err := p.GlobalCleanup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type runResult struct {
+	count         int
+	timeInSeconds float64
+	err           error
+}
+
+func runPerfTest(p NewPerfTest) error {
+	err := runGlobalSetup(p(nil))
+	if err != nil {
+		panic(err)
+	}
+
+	var channels []chan runResult
+	var perfTests []PerfTest
+
+	fmt.Println("=== Test ===")
+
+	w := tabwriter.NewWriter(os.Stdout, 16, 8, 1, ' ', tabwriter.AlignRight)
+	fmt.Fprintln(w, "Current\tTotal\tAverage\t")
+	err = w.Flush()
+	if err != nil {
+		panic(err)
+	}
+
+	for idx := 0; idx < Parallel; idx++ {
+		options := &PerfTestOptions{}
+		if TestProxy != "" {
+			transporter, err := NewProxyTransport(&TransportOptions{TestName: p(nil).GetMetadata().Name})
+			if err != nil {
+				panic(err)
+			}
+			options.ProxyInstance = transporter
+		} else {
+			options.ProxyInstance = defaultHTTPClient
+		}
+		options.ParallelIndex = idx
+
+		perfTest := p(options)
+		perfTests = append(perfTests, perfTest)
+
+		// Run the setup for a single instance
+		err := runSetup(perfTest)
+		if err != nil {
+			return err
+		}
+
+		// Create a thread for running a single test
+		wg.Add(1)
+		c := make(chan runResult, 1) // Create a buffered channel
+		channels = append(channels, c)
+		go runTest(perfTest, c)
+	}
+
+	wg.Wait()
+
+	opsPerSecond := 0.0
+	totalOperations := 0
+	// Get the results from the channels
+	for _, channel := range channels {
+		result := <-channel
+		if err != nil {
+			panic(err)
+		}
+
+		opsPerSecond += float64(result.count) / result.timeInSeconds
+		totalOperations += result.count
+	}
+
+	// Run Cleanup on each parallel instance
+	for _, pTest := range perfTests {
+		err = runCleanup(pTest)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	err = runGlobalCleanup(p(nil))
+
+	fmt.Println("\n=== Results ===")
```

Comment:
It's nice to print results before cleanup, so the user doesn't need to wait if cleanup is slow.